### DISCUSSION TITLE
Notebooks for code example in v0.12

### DISCUSSION THREAD

**Discussion by LeMyst** [LeMyst/WikibaseIntegrator#241] [General]
Hello,

I'll write some notebooks who show how to use the Library in v0.12.
Some are already available at https://github.com/LeMyst/WikibaseIntegrator/tree/rewrite-wbi/notebooks but there is probably more needed.

@pgrond and @dpriskorn maybe you have some ideas ?

Thanks,

Myst~

**Comment by renat-kaufmann** [comment]
Hi Myst!
Nice new syntax, I like it!
1. I see that on write, it returns a lot of data. It might affect performance of load on third party Wikibase instances. What's about making verbose-option or something like that?
2. May be it would make sense to add parallel write option natively? I experimented with concurrent.futures for that and was happy with it, see [here](https://github.com/UB-Mannheim/RaiseWikibase/blob/main/performance_api.py). It would be useful for third-party Wikibase instances.

**Reply by LeMyst** [reply]
Hello @shigapov,

1. write() only return a reference to the object, I don't think there is a real performance impact to return this or nothing.
2. I'll add parallel write to the roadmap and after your question in 1., I think a "batch write" mode (if possible by the API) would be a good approach too, opening the connection to the wikibase instance and waiting for a response are the most time consuming phases.

If you have idea for missing examples in notebooks, don't hesitate to add them to this topic.

Thanks,

**Comment by renat-kaufmann** [comment]
1. Ah, it seems I misunderstood. I thought it returns the whole JSON representation on write (_write method in BaseEntity class), but it returns only QID or PID or LID, right?
2. "Batch write" would be excellent!

Creation of Wikidata properties locally would be useful as well.

**Reply by dpriskorn** [reply]
1) correct 😃
2) +1

**Comment by dpriskorn** [comment]
I'll try looking into the notebooks and see if I can contribute something.

**Comment by pgrond** [comment]
I like the notebooks for documenting and examples!

My use case maybe not the default, but the data I use is for a large part just strings, that should be post processed after import (maybe by hand) to create relevant links between items. During the import I use qualifiers to store those strings as a MonoLinugalText to a property. I also add a reference to track the source and date of import of that data.
So maybe add some examples for adding qualifiers and references.

The main functionality that I'm missing right now in version 0.12 is updating existing items based on a UUID (so without knowing the QID beforehand, but just by sending the data as was possible in 0.11). I believe that is on the roadmap? Can I help with that?
